Jenkinsfile

```diff
@@ -1,69 +1,81 @@
+    parallel failFast: true,
+    "Formatting": {
+        stage("Formatting") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make checkformatting")
             }
-        } finally {
-            postgres.stop()
         }
-
-        nodeEnv = docker.image("node:10-stretch")
-        workspace = pwd()
-        nodeEnv.inside("-u root -e HOME=${workspace}") {
-            sh 'make frontend-tests'
+    },
+    "Docstrings": {
+        stage("Docstrings") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make checkdocstrings")
+            }
         }
-    }
+    },
+    "Backend lint": {
+        stage("Backend lint") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make backend-lint")
+            }
+        }
+    },
+    "Backend tests": {
+        stage("Backend tests") {
+            postgres = docker.image("postgres:9.4").run("-P -e POSTGRES_DB=lmstest")
+            hostIp = sh(script: 'facter ipaddress_eth0', returnStdout: true).trim()
+            containerPort = sh(script: "docker port ${postgres.id} 5432 | cut -d: -f2", returnStdout: true).trim()
+            databaseUrl = "postgresql://postgres@${hostIp}:${containerPort}/lmstest"
 
-    onlyOnMaster {
-        stage('release') {
-            releaseApp(image: img)
+            try {
+                testApp(image: img, runArgs: "${runArgs} -e TEST_DATABASE_URL=${databaseUrl} -e CODECOV_TOKEN=${credentials('LMS_CODECOV_TOKEN')}") {
+                    installDeps()
+                    run("make backend-tests coverage codecov")
+                }
+            } finally {
+                postgres.stop()
+            }
+        }
+    },
+    "Frontend lint": {
+        stage("Frontend lint") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                sh "apk add yarn"
+                run("make frontend-lint")
+            }
+        }
+    },
+    "Frontend tests": {
+        stage("Frontend tests") {
+            workspace = pwd()
+            docker.image("node:10-stretch").inside("${runArgs} -e HOME=${workspace}") {
+                sh "make frontend-tests"
+            }
         }
     }
 }
 
-onlyOnMaster {
-    milestone()
-    stage('qa deploy') {
-        deployApp(image: img, app: 'lms', env: 'qa')
-    }
-
-    milestone()
-    stage('prod deploy') {
-        input(message: "Deploy to prod?")
-        milestone()
-        deployApp(image: img, app: 'lms', env: 'prod')
-    }
+def installDeps() {
+    sh "apk add build-base postgresql-dev python3-dev"
+    sh "pip3 install -q tox>=3.8.0"
 }
 
-def containerPort(container, port) {
-    return sh(
-        script: "docker port ${container.id} ${port} | cut -d: -f2",
-        returnStdout: true
-    ).trim()
+def run(command) {
+    sh "cd /var/lib/lms && ${command}"
```

Comment:
For some reason we seem to need this `cd` before every command we run in the container so I moved it into a function to deduplicate